File: algorithms/leetcode/056-merge-intervals/MergeIntervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        List<int[]> result = new LinkedList<>();
        
        Arrays.sort(intervals, (a, b) -> {
            if (a[0] == b[0]) {
                return a[1] - b[1];
            }
            
            return a[0] - b[0];
        });
        
        for (int i = 0; i < intervals.length; i++) {
            int[] newInterval = new int[]{ intervals[i][0], intervals[i][1] };
            
            while ((i < intervals.length - 1) && intervals[i + 1][0] <= newInterval[1]) {
                newInterval[1] = Math.max(newInterval[1], intervals[i + 1][1]);
                i++;
            }
            
            result.add(newInterval);
        }
        
        return result.toArray(new int[result.size()][2]);
    }
}

File: leetcode/056-merge-intervals/merge-intervals.js
const merge = intervals => {
    const result = [];

    intervals.sort((a, b) => ((a.start - b.start) || (a.end - b.end)));

    for (let i = 0; i < intervals.length; i++) {
        const start = intervals[i].start;
        let end = intervals[i].end;

        while (i < (intervals.length - 1) && (end >= intervals[i + 1].start)) {
            end = Math.max(intervals[i + 1].end, end);
            i++;
        }

        result.push(new Interval(start, end));
    }

    return result;
};
